File: sig-ts/src/core/sign.ts
import { resolve } from 'node:path';
import { sha256, formatHash } from './hash.js';
import { storeSig } from './store.js';
import { logEvent } from './audit.js';
import { loadConfig } from './config.js';
import { createSigContext } from './fs.js';
import { resolveContainedPath } from './paths.js';
import type { SigFS, Signature } from '../types.js';

export interface SignOptions {
  identity?: string;
  engine?: string;
  fs?: SigFS;
}

export async function signFile(
  projectRoot: string,
  filePath: string,
  options?: SignOptions
): Promise<Signature> {
  const ctx = createSigContext(projectRoot, { fs: options?.fs });
  const config = await loadConfig(projectRoot, { fs: ctx.fs });
  const relPath = resolveContainedPath(projectRoot, filePath);
  const absPath = resolve(projectRoot, relPath);
  const content = await ctx.fs.readFile(absPath, 'utf8');
  const hex = sha256(content);

  const identity = options?.identity || config.sign?.identity || defaultIdentity();

  const sig: Signature = {
    file: relPath,
    hash: formatHash(hex),
    algorithm: 'sha256',
    signedBy: identity,
    signedAt: new Date().toISOString(),
    contentLength: Buffer.byteLength(content, 'utf8'),
  };

  const engineName = options?.engine
    || (config.templates?.engine
      ? (Array.isArray(config.templates.engine) ? config.templates.engine[0] : config.templates.engine)
      : undefined);
  if (engineName) {
    sig.templateEngine = engineName;
  }

  await storeSig(ctx, sig, content);
  await logEvent(ctx, {
    event: 'sign',
    file: relPath,
    hash: sig.hash,
    identity: sig.signedBy,
  });

  return sig;
}

function defaultIdentity(): string {
  return process.env.USER || process.env.USERNAME || 'unknown';
}

File: sig-py/src/sig/core/sign.py
from __future__ import annotations

import getpass
import os
from datetime import datetime, timezone
from pathlib import Path

from ..types import Signature
from .hash import sha256, format_hash
from .store import store_sig
from .audit import log_event
from .config import load_config
from .fs import SigFS, create_sig_context
from .paths import resolve_contained_path


def sign_file(
    project_root: str,
    file_path: str,
    *,
    identity: str | None = None,
    engine: str | None = None,
    fs: SigFS | None = None,
) -> Signature:
    ctx = create_sig_context(project_root, fs=fs)
    config = load_config(project_root, fs=ctx.fs)
    rel_path = resolve_contained_path(project_root, file_path)
    abs_path = Path(project_root) / rel_path
    content = ctx.fs.read_file(str(abs_path))
    hex_digest = sha256(content)

    resolved_identity = (
        identity
        or (config.sign.identity if config.sign else None)
        or _default_identity()
    )

    now = datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")

    sig = Signature(
        file=rel_path,
        hash=format_hash(hex_digest),
        algorithm="sha256",
        signed_by=resolved_identity,
        signed_at=now,
        content_length=len(content.encode("utf-8")),
    )

    engine_name = engine
    if engine_name is None and config.templates and config.templates.engine:
        eng = config.templates.engine
        engine_name = eng[0] if isinstance(eng, list) else eng
    if engine_name:
        sig.template_engine = engine_name

    store_sig(ctx, sig, content)
    log_event(
        ctx,
        event="sign",
        file=rel_path,
        hash=sig.hash,
        identity=sig.signed_by,
    )

    return sig


def _default_identity() -> str:
    try:
        return getpass.getuser()
    except Exception:
        return os.environ.get("USER", os.environ.get("USERNAME", "unknown"))
